From 0ac3657d06d27e9685eb74330b325810197a99d6 Mon Sep 17 00:00:00 2001
From: Hauke Mehrtens <hauke.mehrtens@intel.com>
Date: Thu, 19 Nov 2020 18:02:48 +0100
Subject: [PATCH] PONRTSYS-8285: Reinsert NDP and ARP packets

This detects IPv6 NDP and ARP packets and reinserts them on the same
interface where they were received.
These packets are filtered out based on PCE rules by the switch and not
forwarded any more. In some scenarios we want to forward them so they
are also received on the other side of the hardware bridge. To do so
reinsert them into the GSWIP so they can travel through the system as if
there was no PCE rule. If the switch drops them because of some general
traffic forwarding rule this is fine.
---
 drivers/net/datapath/mcast_bridge_filter/Kconfig   |  2 +
 .../mcast_bridge_filter/mcast_bridge_filter.c      | 92 ++++++++++++++++++++++
 2 files changed, 94 insertions(+)

diff --git a/drivers/net/datapath/mcast_bridge_filter/Kconfig b/drivers/net/datapath/mcast_bridge_filter/Kconfig
index 86617687c215..93b78bb5691a 100644
--- a/drivers/net/datapath/mcast_bridge_filter/Kconfig
+++ b/drivers/net/datapath/mcast_bridge_filter/Kconfig
@@ -7,5 +7,7 @@ config DP_MCAST_BRIDGE_FILTER
 	  This installs a netfilter hook which filters out all multicast
 	  packets on all bridges. These packages can then be handled in
 	  user space applications.
+	  ARP and MLD packets on bridges will be reinserted into the
+	  datapath of the GSWIP again.
 
 endif
diff --git a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
index e392a101b6d6..b0cd8fcf6305 100644
--- a/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
+++ b/drivers/net/datapath/mcast_bridge_filter/mcast_bridge_filter.c
@@ -15,7 +15,10 @@
 #include <linux/if_vlan.h>
 #include <linux/in.h>
 #include <linux/ip.h>
+#include <linux/skbuff.h>
 #include <net/addrconf.h>
+#include <net/ndisc.h>
+#include <net/datapath_api.h>
 #include <net/mcast_bridge_filter.h>
 
 /* max number of encapsulated vlan tags */
@@ -29,6 +32,55 @@ void dp_mcast_filter_active(bool active)
 }
 EXPORT_SYMBOL(dp_mcast_filter_active);
 
+/* Reinsert the packet again into the GSWIP Datapath. This way it will flow
+ * through the GSWIP again, but GSWIP will not trap the packet in the PCE rules.
+ * The GSWIP will apply all VLAN forwarding or drop rules like it does for
+ * other packets.
+ */
+static int dp_mcast_reinsert_packet(struct sk_buff *skb)
+{
+	struct dma_tx_desc_0 *dw0;
+	struct dma_tx_desc_1 *dw1;
+	struct sk_buff *nskb;
+	dp_subif_t subif = {0};
+	int ret;
+
+	ret = dp_get_netif_subifid(skb->dev, NULL, NULL, NULL, &subif, 0);
+	if (ret != DP_SUCCESS) {
+		netdev_err(skb->dev, "%s: can not get subif: %d\n", __func__,
+			   ret);
+		return -EIO;
+	}
+
+	nskb = skb_clone(skb, GFP_KERNEL);
+	if (!nskb)
+		return -ENOMEM;
+
+	/* skb->data already points to the beginning of the IP header,
+	 * move it back to the MAC header.
+	 */
+	skb_push(nskb, nskb->data - skb_mac_header(nskb));
+
+	nskb = vlan_hwaccel_push_inside(nskb);
+	if (WARN_ON(!nskb)) {
+		dev_kfree_skb_any(nskb);
+		return -ENOMEM;
+	}
+
+	dw0 = (struct dma_tx_desc_0 *)&nskb->DW0;
+	dw0->field.dest_sub_if_id = subif.subif;
+	dw1 = (struct dma_tx_desc_1 *)&nskb->DW1;
+	dw1->field.ep = subif.port_id;
+
+	ret = dp_xmit(skb->dev, &subif, nskb, nskb->len, DP_TX_INSERT);
+	if (ret) {
+		dev_kfree_skb_any(nskb);
+		netdev_err(skb->dev, "%s: insert fail\n", __func__);
+		return -EIO;
+	}
+	return 0;
+}
+
 /* Iterates over the VLAN tags and increases nhoff for each VLAN tag.
  * returns 0 on success and -EINVAL in case of an parsing error.
  */
@@ -52,6 +104,13 @@ static int dp_mcast_vlan_iter(struct sk_buff *skb, __be16 *proto, int *nhoff)
 	return 0;
 }
 
+static unsigned int dp_mcast_br_hook_arp(struct sk_buff *skb, int nhoff)
+{
+	dp_mcast_reinsert_packet(skb);
+
+	return NF_DROP;
+}
+
 static unsigned int dp_mcast_br_hook_ipv4(struct sk_buff *skb, int nhoff)
 {
 	struct iphdr *iph;
@@ -75,6 +134,32 @@ static unsigned int dp_mcast_br_hook_ipv4(struct sk_buff *skb, int nhoff)
 	return NF_ACCEPT;
 }
 
+/* Returns true for an IPv6 NDP packet.
+ */
+static bool dp_mcast_ipv6_is_ndp_pkt(struct sk_buff *skb, int nexthdr,
+				     int nhoff)
+{
+	struct icmp6hdr *icmp6hdr;
+
+	if (nexthdr != IPPROTO_ICMPV6 ||
+	    !pskb_network_may_pull(skb, nhoff + sizeof(struct icmp6hdr)))
+		return false;
+
+	icmp6hdr = (struct icmp6hdr *)(skb_network_header(skb) + nhoff);
+
+	switch (icmp6hdr->icmp6_type) {
+	case NDISC_ROUTER_SOLICITATION:
+	case NDISC_ROUTER_ADVERTISEMENT:
+	case NDISC_NEIGHBOUR_SOLICITATION:
+	case NDISC_NEIGHBOUR_ADVERTISEMENT:
+	case NDISC_REDIRECT:
+		return true;
+	default:
+		break;
+	}
+	return false;
+}
+
 static unsigned int dp_mcast_br_hook_ipv6(struct sk_buff *skb, int nhoff)
 {
 	const struct ipv6hdr *iph;
@@ -100,6 +185,11 @@ static unsigned int dp_mcast_br_hook_ipv6(struct sk_buff *skb, int nhoff)
 		pr_debug("IPv6 MLD packet dropped");
 		return NF_DROP;
 	}
+
+	if (dp_mcast_ipv6_is_ndp_pkt(skb, nexthdr, nhoff)) {
+		dp_mcast_reinsert_packet(skb);
+		return NF_DROP;
+	}
 	return NF_ACCEPT;
 }
 
@@ -123,6 +213,8 @@ static unsigned int dp_mcast_br_hook(void *priv, struct sk_buff *skb,
 		return dp_mcast_br_hook_ipv4(skb, nhoff);
 	case htons(ETH_P_IPV6):
 		return dp_mcast_br_hook_ipv6(skb, nhoff);
+	case htons(ETH_P_ARP):
+		return dp_mcast_br_hook_arp(skb, nhoff);
 	default:
 		return NF_ACCEPT;
 	}
